### ISSUE TITLE
Validate does not validate tables that reference multiple tables

### ISSUE THREAD

**Issue by mjendruk** [hyrise/hyrise#409]
At the moment, the Validate operator does not validate tables coming from joins and instead passing on the unvalidated table.

**Comment by mrks** [comment]
Same as with sort. Is this an issue? I can't see a reason why you would want to **join** first (which is more expensive if you include unvalidated rows) and validate second. Maybe this should just cause an exception?

**Comment by Bouncner** [comment]
I don't think an exception helps here, if I get the problem right.

In case the anwender writes the query manually, it's her fault. Nothing to do here (assume it's a non-validating OLAP query).
In case we have an SQL query, the query construction needs to ensure that a validation is at latest executed before the join.

I cannot come up with an exact query right now. But there might be problems where we have non-inlinable subqueries that contain joins. If this subquery contains an limit, we might run into problems here. Or do I miss something?

**Comment by mrks** [comment]
Validate removes rows from the result set that logically shouldn't be there. They are only part of it because of how we decided to implement the database. The simplest thing would be to just add a validate behind each GetTable. That's guaranteed to be correct.

Next thing, we could reorder the validate for performance reasons. For projections, it shouldn't make a difference. For table scans, it depends on whether the scan or the validate remove more rows. Joins and sorts are guaranteed to become more expensive with more rows and validate should happen first.

Running a limit on a temp table only works if we validated it before. Otherwise, we don't know how many of the output rows will survive.

Maybe the AST node classes should have a static flag saying if they can deal with unvalidated input.

**Comment by Bouncner** [comment]
I wouldn't say that joins are guaranteed to be faster with validate upfront. Validate introduces an indirection for every qualifying tuple due to emitting a reference chunk and I would guess that >99% usually qualify (some is true for sorts).

In case we have filters (which is pretty much always true) we should validate upfront, but not so much if we don't have a filter (which can be the case with queries as mentioned in my previous post).

Long story short: I am totally for adding the validate per default and letting the optimizer find the optimal position, just as you said. But I don't think we can ignore it or handle it with a simple exception as mentioned yesterday.

**Comment by TwoLaid** [comment]
I like the idea of having a planned `Validate` after each `GetTable` first, and then maybe have the optimizer move the `Validate` up the tree if possible. This will have to keep in mind that validating `Sort` and joins is difficult or impossible.

That should ensure that the results are always correct

**Comment by Bouncner** [comment]
Why should validating the join be hard or impossible (I see that it might be hard to get it fast)?

**Comment by TwoLaid** [comment]
youre right, I jumped the gun there. The `Validate` code looks like it supports multiple referenced tables.

**Comment by mjendruk** [comment]
> I like the idea of having a planned Validate after each GetTable first, and then maybe have the optimizer move the Validate up the tree if possible. 

I like this too.

> The Validate code looks like it supports multiple referenced tables.

At the moment it doesn’t.

**Comment by mrks** [comment]
@Bouncner Yepp, good point. I forgot about the introduction that Validate introduces if it is executed on a stored table.

**Comment by mrks** [comment]
Looking into how MVCC information is currently stored (namely, as three vectors in a chunk), the data concept does not even allow joins to pass the MVCC information to the result. Thus, it does not come as a surprise that none of our joins actually deals with MVCC. Adding this means work not only in the join, but possibly in all other operators. It also means work in the validate operator. All of this is for avoiding one indirection where we don't know if it even makes a significant difference. We probably have more important things to work on.

I propose moving this to the Postponed Issues (if at all) and closing this issue.

**Comment by mrks** [comment]
Since nobody seems to disagree, I moved it to Postponed.

**Comment by mrks** [comment]
This is even more complicated than thought before. Imagine this case:

```
(debug)> print a
=== Columns
|       a||        MVCC        |
|     int||_BEGIN|_END  |_TID  |
=== Chunk 0 ===
|     123||     0|     5|     5|
|    1234||     0|     5|     5|
|   12345||     0|     5|     5|
|       2||     1|     5|     5|
|       2||     7|      |      |
(debug)> print b
=== Columns
|       a||        MVCC        |
|     int||_BEGIN|_END  |_TID  |
=== Chunk 0 ===
|     123||     0|     6|     6|
|    1234||     0|     6|     6|
|   12345||     0|     6|     6|
|       2||     4|     6|     6|
|       2||      |      |     8|
(debug)> txinfo
Active transaction: { transaction id = 8, snapshot commit id = 7 }
```

Think about the query `SELECT * FROM a LEFT JOIN b ON a.a = b.a`. If we don't validate before the join, the join operators wouldn't know if to create NULL values on the right side or not.

A join cannot operate on a non-validated table.

Removed it from postponed.
